client.go

```diff
@@ -31,23 +37,82 @@ func NewClient(endpoint *Endpoint, user, password string) (*Client, error) {
 // NewClientWithParameters will create a new remote client on url, connecting with user and password
 // This function doesn't connect (connection happens only when CreateShell is called)
 func NewClientWithParameters(endpoint *Endpoint, user, password string, params *Parameters) (*Client, error) {
-	transport, err := newTransport(endpoint)
-
-	client := &Client{
-		Parameters: *params,
-		username:   user,
-		password:   password,
-		url:        endpoint.url(),
-		http:       PostRequest,
-		useHTTPS:   endpoint.HTTPS,
-		transport:  transport,
-	}
+	var (
+		client *Client
+		// make custom secure request client auth base
+		secRequest clientAuthRequest
+		// make std clientReques
+		simpleRequest clientRequest
+	)
+
+	// before init everything test if cert and key was passed
+	t := isCertAndKey(endpoint.Cert, endpoint.Key)
+
+	// we should know if the client want's key,cert auth
+	switch t {
+	// secure client auth
+	case true:
+		// make the new custom transport
+		transport, err := newCustomTransport(endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("Can't parse this key and certs: %s", err)
+		}
+		// init the secRequeust with the new transport
+		secRequest.transport = transport
+		// alloc a new client
+		client = &Client{
+			Parameters: *params,
+			username:   user,
+			password:   password,
+			url:        endpoint.url(),
+			http:       secRequest,
+			useHTTPS:   endpoint.HTTPS,
+		}
+		// return the new client
+		return client, nil
 
-	if params.TransportDecorator != nil {
-		client.transport = params.TransportDecorator(transport)
+	// std client request
+	case false:
+		// make new transport that will hold the timeout feature
+		tr, err := newTransport(endpoint)
+		if err != nil {
+			return nil, err
+		}
+		// init the simple request transport
+		simpleRequest.transport = tr
+		// make new client
+		client = &Client{
+			Parameters: *params,
+			username:   user,
+			password:   password,
+			url:        endpoint.url(),
+			useHTTPS:   endpoint.HTTPS,
+		}
+
+		// if do you wish cu custom along more the transport
+		// this feature works very good with pluggble transports because
+		// the custom transports has the same types as in the stdlib
+		if params.TransportDecorator != nil && !t {
+			simpleRequest.transport = params.TransportDecorator(tr)
+		}
+
+		client.http = simpleRequest
+		// returun the client
+		return client, nil
```

Comment:
There's also a lot of code duplication between both cases of the switch. I'm wondering if there couldn't be a better abstraction that would end up in less duplication?